src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+        unsafe {
+            self.reset_overlapped();
+            let buf_ptr = self.read_buf.as_mut_ptr() as LPVOID;
+            let max_read_bytes = buf_cap - buf_len;
+            let ok = kernel32::ReadFile(self.handle,
+                                        buf_ptr.offset(buf_len as isize),
+                                        max_read_bytes as u32,
+                                        &mut bytes_read,
+                                        self.ov_ptr());
+
+            // ReadFile can return TRUE; if it does, an IO completion
+            // packet is still posted to any port, and the OVERLAPPED
+            // structure has the IO operation flagged as complete.
+            if ok == winapi::FALSE {
+                let err = GetLastError();
+                if err == winapi::ERROR_BROKEN_PIPE {
+                    dd!("[$ {:?}:{:?}] BROKEN_PIPE straight from ReadFile", self.iocp, self.handle);
+                    self.closed = true;
+                    return Ok(());
+                }
+
+                if err == winapi::ERROR_IO_PENDING {
+                    self.read_in_progress = true;
+                    return Ok(());
+                }
+
+                Err(WinError::from_system(err, "ReadFile"))
+            } else {
+                self.read_in_progress = true;
+                Ok(())
+            }
+        }
+    }
+
+    fn message_length(&self) -> Option<(usize,usize)> {
+        dd!("[$ {:?}:{:?}] message_length read_buf: {} bytes", self.iocp, self.handle, self.read_buf.len());
+        if self.read_buf.len() < HEADER_SIZE {
+            return None;
+        }
+
+        let buf_header: &[u32] = unsafe { slice::from_raw_parts(self.read_buf.as_ptr() as *const u32, 2) };
+        let data_bytes = buf_header[0] as usize;
+        let oob_bytes = buf_header[1] as usize;
+        let bytes_needed = HEADER_SIZE + data_bytes + oob_bytes;
+        if self.read_buf.len() >= bytes_needed {
+            Some((data_bytes, oob_bytes))
+        } else {
+            None
+        }
+    }
+
+    // Err(false) -> something really failed
+    // Err(true) -> no message
+    // FIXME This is dumb, we should probably make this return Result<Option<(...)>,WinError>
+    // so that we can pass through the error that's lost in map_err below
+    fn get_message(&mut self) -> Result<(Vec<u8>, Vec<OsOpaqueIpcChannel>, Vec<OsIpcSharedMemory>),bool> {
+        let message_lengths = self.message_length();
+        if message_lengths.is_none() {
+            return Err(false)
+        }
+        let (data_bytes, oob_bytes) = message_lengths.unwrap();
+
+        let bytes_needed = HEADER_SIZE + data_bytes as usize + oob_bytes as usize;
+
+        // remove this message's bytes from read_buf, or just take read_buf
+        // if it contains exactly one message
+        let msg_buf = if self.read_buf.len() == bytes_needed {
+            mem::replace(&mut self.read_buf, Vec::with_capacity(READ_BUFFER_SIZE))
```

Comment:
So this is interesting -- `Vec.drain` by itself is faster (~5%) for small sizes, likely due to saving the branch -- but then loses dramatically as the payload size goes up...  especially in the case of out of band channels where every packet will basically saturate the read buffer size.

drain only:

```
test create_channel ... bench:      18,137 ns/iter (+/- 3,548)
test size_00_1      ... bench:       1,876 ns/iter (+/- 591)
test size_01_2      ... bench:       1,835 ns/iter (+/- 518)
test size_02_4      ... bench:       1,886 ns/iter (+/- 540)
test size_03_8      ... bench:       1,809 ns/iter (+/- 319)
test size_04_16     ... bench:       1,818 ns/iter (+/- 328)
test size_05_32     ... bench:       1,842 ns/iter (+/- 268)
test size_06_64     ... bench:       1,914 ns/iter (+/- 425)
test size_07_128    ... bench:       1,980 ns/iter (+/- 443)
test size_08_256    ... bench:       2,151 ns/iter (+/- 296)
test size_09_512    ... bench:       2,422 ns/iter (+/- 456)
test size_10_1k     ... bench:       2,978 ns/iter (+/- 1,594)
test size_11_2k     ... bench:       4,069 ns/iter (+/- 723)
test size_12_4k     ... bench:       5,926 ns/iter (+/- 712)
test size_13_8k     ... bench:       9,481 ns/iter (+/- 1,384)
test size_14_16k    ... bench:      28,817 ns/iter (+/- 5,601)
test size_15_32k    ... bench:      59,549 ns/iter (+/- 19,252)
test size_16_64k    ... bench:      97,283 ns/iter (+/- 28,744)
test size_17_128k   ... bench:     462,368 ns/iter (+/- 106,780)
test size_18_256k   ... bench:     645,383 ns/iter (+/- 231,427)
test size_19_512k   ... bench:   1,504,386 ns/iter (+/- 302,640)
test size_20_1m     ... bench:   2,852,327 ns/iter (+/- 672,506)
test size_21_2m     ... bench:   5,843,816 ns/iter (+/- 1,271,229)
test size_22_4m     ... bench:  11,921,589 ns/iter (+/- 2,496,657)
test size_23_8m     ... bench:  24,080,137 ns/iter (+/- 2,982,565)
```

with mem::replace optimization:

```
test create_channel ... bench:      18,163 ns/iter (+/- 4,020)
test size_00_1      ... bench:       1,998 ns/iter (+/- 424)
test size_01_2      ... bench:       1,978 ns/iter (+/- 197)
test size_02_4      ... bench:       2,009 ns/iter (+/- 349)
test size_03_8      ... bench:       1,987 ns/iter (+/- 566)
test size_04_16     ... bench:       2,029 ns/iter (+/- 208)
test size_05_32     ... bench:       2,028 ns/iter (+/- 313)
test size_06_64     ... bench:       2,026 ns/iter (+/- 375)
test size_07_128    ... bench:       2,002 ns/iter (+/- 327)
test size_08_256    ... bench:       2,044 ns/iter (+/- 495)
test size_09_512    ... bench:       2,250 ns/iter (+/- 558)
test size_10_1k     ... bench:       2,395 ns/iter (+/- 491)
test size_11_2k     ... bench:       2,492 ns/iter (+/- 671)
test size_12_4k     ... bench:       3,134 ns/iter (+/- 694)
test size_13_8k     ... bench:       3,771 ns/iter (+/- 654)
test size_14_16k    ... bench:      17,574 ns/iter (+/- 5,131)
test size_15_32k    ... bench:      22,185 ns/iter (+/- 6,685)
test size_16_64k    ... bench:      33,540 ns/iter (+/- 8,507)
test size_17_128k   ... bench:     380,691 ns/iter (+/- 94,913)
test size_18_256k   ... bench:     418,281 ns/iter (+/- 126,120)
test size_19_512k   ... bench:     961,175 ns/iter (+/- 221,804)
test size_20_1m     ... bench:   1,673,877 ns/iter (+/- 429,866)
test size_21_2m     ... bench:   3,382,619 ns/iter (+/- 787,396)
test size_22_4m     ... bench:   7,339,462 ns/iter (+/- 1,451,900)
test size_23_8m     ... bench:  14,835,211 ns/iter (+/- 2,320,673)
```